Multus metrics

aneeshkp opened the pull request:
Enables metrics to addmission contoller.
Metrics are available at :9091/metrtics

Desc:  Metric to count network attachment definitions.".
Name: network_attachment_definitions_count

Desc:Metric to get total pods using multus
Name: multus_pod_count

Desc:multus_enabled_instance_up (Either 0 or 1)
Name: Metric to identify clusters with multus enabled instances.

dougbtv commented on README.md:
```diff
@@ -60,6 +60,13 @@ EOF
 
+## Collecting metrics with Prometheus
+Network attachment definition admission controller comes with following metrics.
+  1. No of pod's with k8s.v1.cni.cncf.io/networks annotations 
```
Nit: No need to abbreviate, you can use number or `No.` for an abbreviation

aneeshkp replied:
Fixed

dougbtv commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2018 Intel Corporation
```
You can update the date and say `2019 Network Plumbing Working Group`

dougbtv commented on pkg/localmetrics/localmetrics.go:
```diff
@@ -0,0 +1,64 @@
+// Copyright 2019 RedHat
```
`2019 Network Plumbing Working Group`

dougbtv commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,228 @@
+
+const (
+	maxRetries       = 5
+	multusannotation = "k8s.v1.cni.cncf.io/networks"
```
Nit: Mind if we update this and other references to Multus to be generic? E.g. call it `annotationName` or something, as the admission controller itself is generic (doesn't require Multus)

aneeshkp replied:
Fixed

s1061123 commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,225 @@
+	informer  cache.SharedIndexInformer
+}
+
+// Start prepares watchers and run their controllers, then waits for process termination signals
```
Ditto (https://golang.org/doc/effective_go.html#commentary)

aneeshkp commented:
@dougbtv @s1061123 , Added metrics for sriov and multus. Removed CR metrics, it was not accurate unless we introduced controller watching for net-attch-def resource.
 Following metrics are available for telemetry (check metrics doc  linked in readme ) 
The netowrks label will have unique network type names configured for a pod.
any= will give count for all types configured.  
list = will have unique n/w type name
sriov= is special , pick those pod that have sriov type configured
This will give you count for each network type
``` 
network_attachment_definition_instance_total{networks="any"}	5
network_attachment_definition_instance_total{networks="bridge,macvlan"}	2
network_attachment_definition_instance_total{networks="macvlan"}	3 
network_attachment_definition_instance_total{networks="sriov"}	3 
```

and This will indicate whether the cluster has enabled or not for network types
(here you will see one metric per type)

``` 
network_attachment_definition_enabled_instance_up{networks="any"}  1
network_attachment_definition_enabled_instance_up{networks="bridge"} 1
network_attachment_definition_enabled_instance_up{networks="macvlan"}  1
network_attachment_definition_enabled_instance_up{networks="sriov"}  1
``` 

I will be requesting network_attachment_definition_enabled_instance_up and network_attachment_definition_instance_total for telemetry. Hope thats good for this now.
